### PR TITLE
Implement Retry Classifier for Adaptive Rate Limiting

### PR DESCRIPTION
**Pull request by kkashilk**
### Overview

This PR implements the first phase of the Q CLI unified retry strategy, introducing intelligent retry handling with adaptive rate limiting to improve user experience during service overload 
conditions. The implementation leverages AWS Smithy's adaptive retry strategy combined with custom error classification to provide seamless error handling.

#### 🔄 Adaptive Retry Configuration
• **Changed from**: Standard retry mode with single attempt (max_attempts = 1)
• **Changed to**: Adaptive retry mode with client-retry backoff (max_attempts = 3, max_backoff = 10s)
• **Benefits**: Enables client-side rate limiting and automatic throttling detection

#### 🎯 Custom Retry Classifier
• **New file**: crates/chat-cli/src/api_client/retry_classifier.rs
• **Purpose**: Handles Q CLI-specific error patterns and service throttling conditions
• **Key features**:
  • **Monthly Limit Detection**: Identifies MONTHLY_REQUEST_COUNT errors and marks them as RetryForbidden to prevent wasteful retries
  • **Service Overload Handling**: Treats 500/503 errors as ThrottlingError to enable adaptive rate limiting
  • **Priority-based Classification**: Runs after standard classifiers to override decisions for specific error cases

#### 🛡️ Error Classification Strategy

| Error Type | HTTP Status | Classification | Behavior |
|------------|-------------|----------------|----------|
| Monthly Quota Exceeded | 429 + MONTHLY_REQUEST_COUNT | RetryForbidden | No retries |
| Service Overload | 429, 500, 503 | ThrottlingError | Rate limiting + backoff |

#### **Adaptive Retry Benefits**
• **Client-side Rate Limiting**: Automatically activates when throttling errors are detected
• **Token-based Request Gating**: Reduces request rate by 30% immediately upon throttling
• **Exponential Backoff with Jitter**: Prevents thundering herd problems
• **Cubic Recovery Algorithm**: Gradual capacity increase after successful requests

### Testing

• ✅ Unit Tests: Comprehensive test coverage for all error classification scenarios
• ✅ Monthly Limit Detection: Validates MONTHLY_REQUEST_COUNT error handling
• ✅ Service Overload Detection: Tests 500/503 status code classification
• ✅ Priority Verification: Ensures classifier runs with correct priority
• ✅ Edge Cases: Handles missing response bodies and malformed responses

#### **Before**
• Single retry attempt with immediate failure on service overload
• No differentiation between retryable and non-retryable errors
• Users manually retry requests during service issues

#### **After**
• Intelligent retry with up to 3 attempts for retryable errors
• Automatic rate limiting prevents overwhelming overloaded services
• Monthly quota errors fail fast without wasteful retries
• Transparent handling of service overload conditions

### Implementation Roadmap

This PR represents Phase 1 of the unified retry strategy:

• ✅ Phase 1: Core retry infrastructure with custom error classification
• 🔄 Phase 2: Delay tracking interceptor with user notifications (next PR)
• 🔄 Phase 3: Production rollout with telemetry and monitoring

#### **Backend Protection**
• Client-side rate limiter prevents retry storms
• Maximum attempt limits prevent infinite retry loops
• Custom classifier can mark errors as RetryForbidden for immediate failure

#### **Backward Compatibility**
• No breaking changes to existing API
• Graceful degradation if retry classifier fails
• Maintains existing timeout and stalled stream protection

### Related Documentation

• 🔗 AWS Smithy Retry Documentation: [Adaptive Retry Strategy](https://docs.rs/aws-smithy-runtime-api/latest/aws_smithy_runtime_api/client/retries/)

### CODE REVIEW

File: crates/chat-cli/src/api_client/retry_classifier.rs [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+    RetryClassifierPriority,
+};
+use tracing::debug;
+
+/// Error marker for monthly limit exceeded errors
+const MONTHLY_LIMIT_ERROR_MARKER: &str = "MONTHLY_REQUEST_COUNT";
+
+/// Error message for high load conditions that should be retried
+const HIGH_LOAD_ERROR_MESSAGE: &str =
+    "Encountered unexpectedly high load when processing the request, please try again.";
```
Comment (kensave) [inline crates/chat-cli/src/api_client/retry_classifier.rs:16]: How deterministic are these? What happens if the message changes in back-end.

Comment (kkashilk) [reply crates/chat-cli/src/api_client/retry_classifier.rs:16]: These will be deterministic, and we'll generally ensure backwards compatibility.

We already have a similar check in the api-client module - https://github.com/aws/amazon-q-developer-cli/blob/main/crates/chat-cli/src/api_client/mod.rs#L386-L408

Comment (brandonskiser) [reply crates/chat-cli/src/api_client/retry_classifier.rs:16]: Between this file and the linked one, there is a lot of complicated logic + duplication going on now... I would really like to consolidate a lot of these checks.

We have the following modeled errors:
1. Model overloaded error (either 429 throttling or 500 w/ specific message check), returned as `ApiClientError::ModelOverloadedError`
2. General throttling error (just 429), returned as `ApiClientError::QuotaBreach`
3. Monthly limit reached (just a message check), returned as  `ApiClientError::MonthlyLimitReached`

Can we consolidate the logic for all of these checks? Just parameterize on the status code and request body probably. We could add an additional like service unavailable one for 503

Comment (kkashilk) [reply crates/chat-cli/src/api_client/retry_classifier.rs:16]: yeah, that makes sense, will update

**Review by brandonskiser — APPROVED** [review]

**Review by kensave — APPROVED** [review]
